cosigner/mfa/cosigner.go

```diff
@@ -0,0 +1,66 @@
+package mfa
+
+import (
+	"crypto"
+	"encoding/json"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/openpubkey/openpubkey/pktoken"
+)
+
+type Cosigner struct {
+	csid string
+	kid  string
+
+	alg    jwa.KeyAlgorithm
+	signer crypto.Signer
+
+	// The MFA in MFACosigner, we use this to authenticate the user
+	mfa Authenticator
+}
+
+type Authenticator interface {
+	Authenticate(pkt *pktoken.PKToken) error
+	URI() string
+}
+
+func NewCosigner(signer crypto.Signer, alg jwa.SignatureAlgorithm, csid, kid string, authenticator Authenticator) (*Cosigner, error) {
+	return &Cosigner{
+		csid:   csid,
+		kid:    kid,
+		alg:    alg,
+		signer: signer,
+		mfa:    authenticator,
+	}, nil
+}
+
+func (c *Cosigner) Cosign(pkt *pktoken.PKToken) error {
+	if err := c.mfa.Authenticate(pkt); err != nil {
+		return err
+	}
+
+	protected := pktoken.CosignerClaims{
+		ID:          c.csid,
+		KeyID:       c.kid,
+		Algorithm:   c.alg.String(),
+		AuthID:      "12345678",
```

Comment:
AuthID in the OpenPubkey paper is an auth session identifier so that if the MFA Cosigner logged details about when it granted the signature, the signature/cosigner claims in a PK Token could be linked back to a specific time and place.

This enables the following features:

* A user could request that the Cosigner revoke a particular auth session/PK Token, much like users in gmail can revoke particular gmail sessions.

* A forensic investigator or incident response, could track a compromise and say see what MFA device was used to authenticate to the MFA Cosigner and determine which MFA devices were compromised.

* Risk and Anomaly based revocation of particular PK Tokens associated with particular sessions

Reply:
@jonnystoten you do ask a good question. I would like to revisit all the claims at some point in the future when the architecture and POP Auth have been implemented so we can better understand what's necessary. For now, I implemented them verbatim from the paper.

Reply:
If you don't mind @lgmugnier I might address this issue in a PR that builds on this one, after this is merged.